### PR TITLE
Discuss seq parsing

### PR DESCRIPTION
**Pull request by HeroicKatora**
This changes the parser to accept arguments to short flags again. The root of the problem is clap's overly restrictive parsing code, specifically this line:

https://github.com/clap-rs/clap/blob/5c3868ea4cb8063731d8526e8e97414942a987ae/src/parse/parser.rs#L994-L995

This means that `-s,` is not accepted when `AllowHyphenValues` is on, as it contains `,` which isn't a short option flag. This is despite the short option taking an argument. As we can see from the surrounding code the treatment is very different if we merely activate `AllowNegativeNumbers`. This, still, isn't _exactly_ the same parser style as GNU `seq`. In particular, that program matches value arguments by the expression: `-[0-9.].*` (regex style) instead. This implies that the _error message_ for negative non-numbers, but also for negative numbers whose format isn't recognized by the standard `<f64 as FromStr>` impl, such as `-0x.ep-3` differs.

In short, I believe that solving both #2729 and #2660 isn't reliably possible with the current `clap` version.

I've tried a hacky way, however judged the resulting code as _far_ too complex as we need to semantically know which arguments are option values, i.e. it was more like a fully customized parser than a short hack..

We don't need to accept this PR. This is mostly to summarize my findings on my weekend journey on #2729 . We can escalate this problem to clap, to create some parser settings and version where the parser works as expected. Or we can do nothing on this issue and prioritize the number formats of #2660 which require `AllowHyphenValues`.

Closes: #2729

### CODE REVIEW

**Comment by tertsdiepraam** [comment]
Excellent work! Thank you for all this research and experimentation!

If I understand correctly, we either need a custom parser or clap needs to be able to support this? Do you see any specific way that clap can be updated in a backwards compatible way to support our use case? They (rightfully) adhere strictly to semver.

For instance, we could ask clap to:
- support an extended floating-point number syntax (unlikely that they'll accept that),
- include a flag that gives short options priority over hyphen values, or
- support some custom check for which hyphen values to allow, possibly by regex or a function returning a bool.

These are all quite complex, so is there anything simpler we could propose? If not, then we probably have to use a custom parser. I suggest we then filter out the numbers and then pass the rest of the options to clap (without `AllowHyphenValues` or `AllowNegativeNumbers`). Is that the approach that you tried or did you try it some other way?

**Comment by HeroicKatora** [comment]
The simplest option we might propose to `clap` might be an extension of `AllowHyphenValues`, where the match against short options is more permissive. In particular, the logic could be: when encountering any short option taking a value, assume the rest of the argument is that value (instead of further short options as it happens now).

The filter approach was what I tried and it works somewhat poorly. In particular, to identify numbers we already need to semantically identify all arguments that are options and all that are values, since we must only filter out the latter! This is already a full parser that needs to be written and it must be semantically specific to the short option characters. That's … quite complex for an ad-hoc hack.

A more complex addition to clap might be one that allows us to write an option parser that works more similar to what GNU does in `getlong_opt`. That is, where we _poll_ the parser for one option at a time but stay in control of the loop. This is what actually allows GNU `seq` to decide on its own which argument constitutes a floating point value (and thus stop option parsing). See https://github.com/coreutils/coreutils/blob/master/src/seq.c#L593-L604

The basic capabitilities for such an extension are already present in `clap`. In [`try_get_matches_from_mut`](https://docs.rs/clap/3.0.14/clap/struct.App.html#method.try_get_matches_from_mut) the original app is preserved. What's missing is a method that consumes only _one_ argument into some result structure instead of all of them so that we can routinely peek into the remaining arguments.

**Comment by tertsdiepraam** [comment]
> The simplest option we might propose to `clap` might be an extension of `AllowHyphenValues`, where the match against short options is more permissive.

All right! That's what I was proposing in my second point. I think this is a viable solution!

> A more complex addition to clap might be one that allows us to write an option parser that works more similar to what GNU does in getlong_opt. That is, where we poll the parser for one option at a time but stay in control of the loop. This is what actually allows GNU seq to decide on its own which argument constitutes a floating point value (and thus stop option parsing).

That's an interesting solution. I wonder whether they would be in favor of a more "low-level" API like that.

There are also (long-term) plans to make clap more modular and allow for custom parsers. The discussion around that is here: https://github.com/clap-rs/clap/issues/2915. If they eventually do so, we can probably make a more GNU compatible parser and apply it to all utils.

I think there's no point to keep guessing here. Could you open an issue on clap with the solution you like best?

**Comment by rivy** [comment]
* Rebased to 'main' to check updated GnuTests CI workflow.

**Review by tertsdiepraam — APPROVED** [review]
